### ISSUE TITLE
Bad results when training hacnn model with triplet loss engine.

### ISSUE THREAD

**Issue by Rajat-Mehta** [KaiyangZhou/deep-person-reid#191]
I tried to train hacnn model with triplet loss on my custom dataset (VeRi dataset) and got very bad results. Here are the training settings that I used:

```
from dataset_torchreid import *
import argparse


parser = argparse.ArgumentParser(description='Training torchreid')
parser.add_argument('--engine', default='softmax', type=str, help='which engine to use: softmax or triplet')
opt = parser.parse_args()

engine = opt.engine

torchreid.data.register_image_dataset('veri_dataset', VeRiDataset)

datamanager = torchreid.data.ImageDataManager(
    root='..Dataset/VeRi_with_plate/',
    sources='veri_dataset',
    height=160,
    width=64,
    train_sampler='RandomIdentitySampler' if engine == "triplet" else "RandomSampler"
)


model = torchreid.models.build_model(
    name='hacnn',
    num_classes=datamanager.num_train_pids,
    loss=engine
)

model = model.cuda()

optimizer = torchreid.optim.build_optimizer(
    model, optim='adam', lr=0.0003
)

scheduler = torchreid.optim.build_lr_scheduler(
    optimizer,
    lr_scheduler='single_step',
    stepsize=20
)

if engine == "triplet":
    save_path = './model/veri_hacnn/hacnn_triplet/'
    model_engine = torchreid.engine.ImageTripletEngine(
        datamanager, model, optimizer, margin=0.3,
        weight_t=1, weight_x=0, scheduler=scheduler
    )
else:
    save_path = './model/veri_hacnn/hacnn_softmax/'
    model_engine = torchreid.engine.ImageSoftmaxEngine(
        datamanager, model, optimizer, scheduler=scheduler
    )

model_engine.run(
    max_epoch=100,
    save_dir=save_path,
    print_freq=10
)
```

Here are the results that I got after training the model for 100 epochs:
```
Successfully loaded pretrained weights from "/home/rajat/MyPC/DFKI/MasterThesis/Vehicle_Reidentification/model/veri_hacnn/hacnn_triplet/model.pth.tar-100"
##### Evaluating veri_dataset (source) #####
Extracting features from query set ...
/home/rajat/MyPC/DFKI/MasterThesis/Vehicle_Reidentification/veri/lib/python3.6/site-packages/torch/nn/functional.py:2457: UserWarning: nn.functional.upsample is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample is deprecated. Use nn.functional.interpolate instead.")
Done, obtained 1678-by-1024 matrix
Extracting features from gallery set ...
Done, obtained 11579-by-1024 matrix
Speed: 0.0614 sec/batch
Computing distance matrix with metric=euclidean ...
Computing CMC and mAP ...
** Results **
mAP: 9.6%
CMC curve
Rank-1  : 14.7%
Rank-5  : 25.9%
Rank-10 : 34.9%
Rank-20 : 43.4%
Visualizing top-9 ranks
# query: 1678
# gallery 11579
Saving images to "./model/veri_hacnn/hacnn_triplet/visrank-1/veri_dataset"
Done
```

If I train the same model but with softmax engine I get pretty good results as compared. I get 60% mAP with softmax engine. But as can be seen in the results above, after training for about 9 hours and for 100 epochs(with triplet loss engine) the mAP came out to be just 9.6%. What could have gone wrong in the training process?

Thanks in Advance.

**Comment by Rajat-Mehta** [comment]
I think the bad performance is because of the size transformation of the images. As HACNN network is designed specifically for person reidentification task, it accepts images of size 160 * 64 only. So, when the images of vehicles are resized to 160 * 64 the vehicles become completely distorted and thus leading to bad training. 

@KaiyangZhou, Could you please verify this or do you think there is some other reason that might be causing this bad performance of HACNN on VeRi(vehicle) datasets ?

**Comment by KaiyangZhou** [comment]
triplet loss does not work well alone for training a model from scratch, I would suggest using softmax only which guarantees performance

I would not use strip-based models (e.g. hacnn) for vehicle reid as vehicles can be upside down

how about using resnet50 or densenet121?

I don't think resizing image is the culprit, the image textures are more important (this paper also discusses this point https://openreview.net/forum?id=Bygh9j09KX)

**Comment by joaqo** [comment]
@Rajat-Mehta Hi, would it be at all possible for you to share your custom script for adding VeRi as a dataset? Thanks!

**Comment by Rajat-Mehta** [comment]
@joaqo you can find my scipts here: https://github.com/Rajat-Mehta/Vehicle_Reidentification/blob/master/dataset_torchreid.py

**Comment by joaqo** [comment]
Thanks a lot! @Rajat-Mehta
